### ISSUE TITLE
List of services and actions not working for me

### ISSUE THREAD

**Issue by vforv** [hemerajs/hemera-cli#2]
here is my code:

http://pushsc.com/show/code/58f8c1ee0a58d602a0b99d23

connected to nats server like this:
`./gnatsd`
and get like this:
`Server is ready`
and run code above... 

Here is image for hemera:
http://i.imgur.com/aV1H5YH.png

Always get no services available....

**EDIT:**

I made it working for `actions`(forgot hemera-stats) but have problem for `services`

When I type services it just freez...

**Comment by StarpTech** [comment]
Hi @vforv as mentioned in the readme every hemera service has to use the `hemera-stats` plugin to expose the interface.

```js
hemera.use(require('hemera-stats'))
```

**Comment by vforv** [comment]
I added `hemera-stats` it now working for actions but not working for services, it just freez when I type: `services`...

**Comment by StarpTech** [comment]
You got no output? Could you post your code?

**Comment by vforv** [comment]
Yes, it freez so I need to type Ctrl + C to exit console...

Here is the code:

```
'use strict'

const Hemera = require('./../packages/hemera')
const HemeraStats = require('hemera-stats')
const nats = require('nats').connect()
const HemeraJoi = require('hemera-joi')

const hemera = new Hemera(nats, { logLevel: 'info' })
hemera.use(HemeraJoi)
const Hapi = require('hapi')

const server = new Hapi.Server()
server.connection()
hemera.use(HemeraStats)

server.register({
  register: require('hapi-hemera'),
  options: {
    hemera:{
      logLevel: 'info' 
    },
    nats: 'nats://localhost:4222'
  }
}, function(err) {
    server.start((err) => {

    if (err) {
      throw err;
    }
    console.log(`Server running at: ${server.info.uri}`);
    
   })
})

hemera.ready(() => {

  let Joi = hemera.exposition['hemera-joi'].joi

  hemera.add({ 
    topic: 'math',
    cmd: 'add',
    a: Joi.number().required(),
    b: Joi.number().required()
  }, (req, cb) => {
    cb(null, parseInt(req.a) + parseInt(req.b))
  })

  server.route({
    method: 'GET',
    path: '/api/add',
    handler: function (request, reply) {

      hemera.act({
          topic: 'math',
          cmd: 'add',
          a: request.query.a,
          b: request.query.b,
          refresh: !!request.query.refresh
        },
        (err, result) => {

          if (err) {
            return reply(Boom.wrap(err.cause, 400))
          }

          return reply(result)
        })
    }
  })
  
    
})
```

**Comment by StarpTech** [comment]
Do you run this example from the repository? 

`const Hemera = require('./../packages/hemera')`

**Comment by vforv** [comment]
Yes i cloned `hemera` folder installed deps and changed /examples/server.js and run it...

**Comment by StarpTech** [comment]
Hi @vforv I got it. You dont have to create a hemera instance manually. The plugin do it all for you. 
Please upgrade to `+ hapi-hemera@0.1.3` and use this snippet [Hemera with Hapi](https://gist.github.com/StarpTech/94c2c1591fbb16651b526ac0add52692).

```
curl http://127.0.0.1/api/add?a=10&b=20
```

**Comment by vforv** [comment]
Aha ok, I run the script but still have same problem with `services` on hemera-cli...

**Comment by StarpTech** [comment]
@vforv it works fine on my machine. NATS is running?

**Comment by vforv** [comment]
Yes, check screenshot:

http://dodaj.rs/photos/20170422149288619162064.png

for actions works but for services not

This are my steps:
1. connect to nats server 
2. c/p script you send me and run it 
3. login to hemera-cli, connect to server, run services

Any additional step I need?

**Comment by StarpTech** [comment]
Hi @vforv you also have to use [hemera-stats](https://github.com/hemerajs/hemera/tree/master/packages/hemera-stats) plugin to expose your service for the `services` call.

**Comment by augustine-tran** [comment]
The same for me. I can see list actions but list services doesn't work.

**Comment by StarpTech** [comment]
@rualatngua which versions are you using?

**Comment by StarpTech** [comment]
Please upgrade hapi-hemera, hemera-stats and hemera-cli to the latest version.

**Comment by majest** [comment]
Unfortunately that doesn't work for me as well. actions is fine, services hangs the cli. 

```
"dependencies": {
    "hemera-stats": "^1.0.22",
    "hemera-zipkin": "^1.0.21",
    "nats": "^0.7.20",
    "nats-hemera": "^1.3.18"
  }
```

hemera-cli :  `~/.npm/hemera-cli/1.0.11`

index.js 

```
...
hemera.ready(() => {
...

    hemera.act({
      topic: 'stats',
      cmd: 'processInfo'
    }, function (err, resp) {
      console.log(err, resp)
    })

    hemera.act({
      topic: 'stats',
      cmd: 'registeredActions'
    }, function (err, resp) {
      console.log(err, resp)
    })
})
```

**Comment by StarpTech** [comment]
Hi @majest you also use the `hemera-stats` plugin in your services to expose that interface? `hemera-stats` will expose the service in the network to receive information about it.

**Comment by StarpTech** [comment]
1. Run `examples/monitoring/stats.js`
2. Install `hemera-cli` globally
3. `$ hemera-cli`

Output:

Services:
```
E:\Repositorys\hemera [master ≡ +0 ~1 -0 !]> hemera-cli
Welcome to the Hemera CLI! v1.0.11
local@starptech~$ connect
NATS Server Url: nats://localhost:4222
Username:
Password:
Connected to NATS Server
local@starptech~$ services


┌───────────────────────────────────────────────────┬────────────────┬─────────────┬───────────┬──────────┬──────────┬──────────────────────────┐
│ Service                                           │ Uptime         │ Env         │ LoopDelay │ Heap     │ Rss      │ Date                     │
├───────────────────────────────────────────────────┼────────────────┼─────────────┼───────────┼──────────┼──────────┼──────────────────────────┤
│ hemera-starptech-a6eaf697ee574d4087af5037f6986874 │ 44 seconds ago │ development │ 0.10ms    │ 14.85 MB │ 35.89 MB │ 2017-07-18T11:16:43.273Z │
└───────────────────────────────────────────────────┴────────────────┴─────────────┴───────────┴──────────┴──────────┴──────────────────────────┘


local@starptech~$
```
Actions:
```
┌───────────────────────────────────┬───────────────────────────────────────────────────┬──────────────┐
│ Pattern                           │ Service                                           │ Plugin       │
├───────────────────────────────────┼───────────────────────────────────────────────────┼──────────────┤
│ cmd:processInfo,topic:stats       │ hemera-starptech-a6eaf697ee574d4087af5037f6986874 │ hemera-stats │
├───────────────────────────────────┼───────────────────────────────────────────────────┼──────────────┤
│ cmd:registeredActions,topic:stats │ hemera-starptech-a6eaf697ee574d4087af5037f6986874 │ hemera-stats │
├───────────────────────────────────┼───────────────────────────────────────────────────┼──────────────┤
│ cmd:add,topic:math                │ hemera-starptech-a6eaf697ee574d4087af5037f6986874 │ core         │
└───────────────────────────────────┴───────────────────────────────────────────────────┴──────────────┘
```

**Comment by StarpTech** [comment]
Please reopen when issue could not be fixed.

**Comment by Luke-Tang** [comment]
@StarpTech  same for me. actions is fine but services is not.

```
 "dependencies": {
    "hemera-joi": "^1.2.1",
    "hemera-stats": "^1.0.23",
    "nats": "^0.7.20",
    "nats-hemera": "^1.5.1"
  }
```


```
const Hemera = require("nats-hemera")
const HemeraStats = require('hemera-stats')
const nats = require("nats").connect()
const HemeraJoi = require('hemera-joi')

const hemera = new Hemera(nats, {
    logLevel: "info",
    load: {
        process:{
            sampleInterval:100
        }
    }
})

hemera.use(HemeraStats)
hemera.use(HemeraJoi)

hemera.ready(function () {

    hemera.add({ topic: "math", cmd: "add" }, function (resp, cb) {

        cb(null, resp.a + resp.b)
    })

    hemera.act({ topic: "math", cmd: "add", a: 1, b: 2 }, function (err, resp) {

        console.log("Result", resp)
    })
})
```

```
tangkundeMBP:device-center tangkun$ hemera-cli
Welcome to the Hemera CLI! v1.1.0
local@tangkundeMacBook-Pro~$ connect
NATS Server Url: nats://localhost:4222
Username: 
Password: 
Connected to NATS Server
local@tangkundeMacBook-Pro~$ actions


┌───────────────────────────────────┬────────────────────────────────────────────────────────────────────┬──────────────┐
│ Pattern                           │ Service                                                            │ Plugin       │
├───────────────────────────────────┼────────────────────────────────────────────────────────────────────┼──────────────┤
│ cmd:processInfo,topic:stats       │ hemera-tangkundeMacBook-Pro.local-213e1a94883d4539b84852eb4109a056 │ hemera-stats │
├───────────────────────────────────┼────────────────────────────────────────────────────────────────────┼──────────────┤
│ cmd:registeredActions,topic:stats │ hemera-tangkundeMacBook-Pro.local-213e1a94883d4539b84852eb4109a056 │ hemera-stats │
├───────────────────────────────────┼────────────────────────────────────────────────────────────────────┼──────────────┤
│ cmd:add,topic:math                │ hemera-tangkundeMacBook-Pro.local-213e1a94883d4539b84852eb4109a056 │ core         │
└───────────────────────────────────┴────────────────────────────────────────────────────────────────────┴──────────────┘


local@tangkundeMacBook-Pro~$ services


```

**Comment by StarpTech** [comment]
Hi @Luke-Tang I copy & paste your example and it works I cant reproduce it.

```
local@starptech~$ actions





┌───────────────────────────────────────────────────┬────────────────┬─────────────┬───────────┬──────────┬──────────┬──────────────────────────┐
│ Service                                           │ Uptime         │ Env         │ LoopDelay │ Heap     │ Rss      │ Date                     │
├───────────────────────────────────────────────────┼────────────────┼─────────────┼───────────┼──────────┼──────────┼──────────────────────────┤
│ hemera-starptech-9955820ce3a24aee83c78d1cf403ec30 │ 25 seconds ago │ development │ 0.68ms    │ 12.79 MB │ 35.41 MB │ 2017-08-02T09:52:00.839Z │
└───────────────────────────────────────────────────┴────────────────┴─────────────┴───────────┴──────────┴──────────┴──────────────────────────┘
```

**Comment by StarpTech** [comment]
Hi @Luke-Tang I think its a cross platform issue. I released a potential fix please try hemera-cli 1.1.2 thanks!

**Comment by Luke-Tang** [comment]
Hi @StarpTech  hemera-cli@1.1.2 and  hemera-cli@1.1.4 are both not work.  I tested it on mac and pc. so, it may not be a cross platform issue.

**Comment by Luke-Tang** [comment]
Hi @StarpTech I think it's a Env issue.   I clone and trace this code.   I found the value of service.nodeEnv(line 167 in index.js) is undefined.

I delete 'Env' and 'service.nodeEnv' from code

```
      const table = new CliTable({
        head: ['Service', 'Uptime', /*'Env',/* 'LoopDelay', 'Heap', 'Rss', 'Date']
      })

      ......

      _.each(services, (service) => {
        table.push([
          service.app,
          Humanize.relativeTime(Humanize.time() - service.uptime),
          /*service.nodeEnv,*/
          Humanize.numberFormat(service.eventLoopDelay) + 'ms',
          Humanize.filesize(service.heapUsed),
          Humanize.filesize(service.rss),
          new Date(service.ts).toISOString()
        ])
      }) 
```

```
┌─────────────────────────────────────────────────┬───────────────────┬───────────┬──────────┬──────────┬──────────────────────────┐
│ Service                                         │ Uptime            │ LoopDelay │ Heap     │ Rss      │ Date                     │
├─────────────────────────────────────────────────┼───────────────────┼───────────┼──────────┼──────────┼──────────────────────────┤
│ hemera-Luke-PC-93c309287a824195a6c087e6095eaa5a │ about an hour ago │ 0.33ms    │ 11.19 MB │ 19.33 MB │ 2017-08-03T07:37:27.044Z │
└─────────────────────────────────────────────────┴───────────────────┴───────────┴──────────┴──────────┴──────────────────────────┘
```

services cmd works

**Comment by StarpTech** [comment]
@Luke-Tang that makes sense thanks for the fix I released it with 1.1.5
